projecttacoma/fqm-execution
# Implement library data requirements

# Summary
Before, we had only the option to calculate dataRequirements on a measure bundle. Now, we can calculate dataRequirements on a collection of Library resources (library bundle). This will be helpful when we implement Library/$data-requirements in the measure-repository-service.

## New behavior
New function in `Calculator.ts` calculates library dataRequirements from a Library bundle. This is in addition to the calculateDataRequirements function that calculates the dataRequirements on a measure bundle.

## Code changes
- `src/calculation/Calculator.ts` - new function `calculateLibrarydataRequirements` gets data requirements for a collection of Library resources. A lot of this code was identical to that in `calculateDataRequirements` but we wanted them to be separate functions so I refactored the duplicate code into a function in `DataRequirementHelpers.ts`.
- `src/cli.ts` - new cli option to run `libraryDataRequirements` with a flag with the `rootLibRef`.
- `src/helpers/DataRequirementHelpers.ts` - refactored a lot of the identical code in `calculateDataRequirements` and `calculateLibraryDataRequirements` to the `getDataRequirements` function.
- `src/helpers/MeasureBundleHelpers.ts` - `extractLibrariesFromLibraryBundle` gets libraries from a library bundle and `extractLibrariesFromMeasureBundle` gets libraries from a measure bundle. They both use `extractLibrariesFromBundle` because a lot of the code was the same. The main difference is that `extractLibrariesFromLibraryBundle` has to take a reference to the root library as a parameter, since there is no measure on the bundle to indicate the root library like in `extractLibrariesFromMeasureBundle`.
- `src/helpers/elm/ELMInfoCache.ts` - changed `extractLibrariesFromBundle` to `extractLibrariesFromMeasureBundle`.
- `src/types/Calculator.ts` - added `rootLibRef` as an optional CalculationOption which we need in `libraryDataRequirements` in the cli.
- `test/unit/helpers/ELMInfoCache.test.ts` - changed `extractLibrariesFromBundle` to `extractLibrariesFromMeasureBundle`.
- `test/unit/helpers/MeasureBundleHelpers.test.ts` - added a test for `extractLibrariesFromLibraryBundle`. This uses the `ELM130-7.3.000-bundles-nocodes.json` fixture that we already have but filters is so that it is a bundle of only its library resources. 

# Testing guidance
- `npm run check`
- Try the new cli command! 
- My motivation behind this test is I want to take a measure bundle and run dataRequirements on it. Then, I want to take that same measure bundle, but remove all resources that are not of type Library. Finally, I run libraryDataRequirements on that library bundle and then compare the two outputs. They should be the same.
- With this approach, I used the `EXM130-7.3.000-bundle-nocodes.json` fixture because it is there. I suggest you do the same since I attached that file and the one with only the library resources below, but feel free to try other bundles!
- `npm run cli -- dataRequirements -m <path-to-EXM130-7.3.000-bundle-nocodes.json> -o` 
- Save that `output.json` file
- `npm run cli -- libraryDataRequirements -m <path-to-EXM130.7.3.000-library-nocodes.json --root-lib-ref "Library-library-EXM130-7.3.000 -o`
- Save that `output.json` file.
- Compare the two files and make sure they are the same!


[LibraryDataRequirementsTestFiles.zip](https://github.com/projecttacoma/fqm-execution/files/10362517/LibraryDataRequirementsTestFiles.zip)
